File: GrowthWare_CS/GrowthWareFrameworkWeb/Utilities/ClientChoicesUtility.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using GrowthWare.Framework.BusinessLogicLayer;
using GrowthWare.Framework.Model.Profiles;
using GrowthWare.Framework.Web.Controllers;
using System.Web;
using System.Globalization;

namespace GrowthWare.Framework.Web.Utilities
{
	/// <summary>
	/// ClientChoicesUtility serves as the focal point for any web application needing to utiltize the GrowthWare framework
	/// with regards to ClientChoices.
	/// </summary>
	public class ClientChoicesUtility
	{
		private MSecurityEntityProfile m_SecurityEntityProfile = null;
		private BClientChoices m_BClientChoices = null;
		private String m_CachedAnonymousChoicesState = "AnonymousClientChoicesState";

		/// <summary>
		/// Constructor sets up the private fields
		/// </summary>
		public ClientChoicesUtility() 
		{ 
			m_SecurityEntityProfile = SecurityEntityUtility.GetDefaultProfile();
			m_BClientChoices = new BClientChoices(ref m_SecurityEntityProfile, WebConfigSettings.CentralManagement);
		}

		/// <summary>
		/// Returns the client choices given the account
		/// </summary>
		/// <param name="account">String</param>
		/// <returns>MClientChoicesState</returns>
		public MClientChoicesState GetClientChoicesState(String account) 
		{ 
			MClientChoicesState mRetVal = null;
			if(HttpContext.Current.Session != null)
			{
				mRetVal = (MClientChoicesState)(HttpContext.Current.Session[MClientChoices.SessionName]);
			}
			if(mRetVal == null)
			{
				if (account.Trim().ToLower(CultureInfo.CurrentCulture) == "anonymous")
				{
					mRetVal = (MClientChoicesState)HttpContext.Current.Cache[m_CachedAnonymousChoicesState];
					if (mRetVal == null)
					{
						mRetVal = m_BClientChoices.GetClientChoicesState(account);
						CacheController.AddToCacheDependency(m_CachedAnonymousChoicesState, mRetVal);
					}
				}
				else 
				{
					if (mRetVal.AccountName != account)
					{
						mRetVal = m_BClientChoices.GetClientChoicesState(account);
					}
				}
			}
			if(HttpContext.Current.Session != null)
			{
				HttpContext.Current.Session.Add(MClientChoices.SessionName, mRetVal);
			}
			return mRetVal;
		}

		/// <summary>
		/// Save the client choices to the database.
		/// </summary>
		/// <param name="clientChoicesState">MClientChoicesState</param>
		/// <remarks></remarks>
		public void Save(ref MClientChoicesState clientChoicesState) 
		{ 
			m_BClientChoices.Save(ref clientChoicesState);
			if(HttpContext.Current.Session != null)
			{
				HttpContext.Current.Items[MClientChoices.SessionName] = clientChoicesState;
			}
		}
	}
}

File: GrowthWare_VB/GrowthWareFrameworkWeb/Utilities/ClientChoicesUtility.vb
Imports System.Globalization
Imports System.Web
Imports GrowthWare.Framework.BusinessLogicLayer
Imports GrowthWare.Framework.Model.Profiles
Imports GrowthWare.Framework.Web.Controllers

Namespace Utilities
	''' <summary>
	''' ClientChoicesUtility serves as the focal point for any web application needing to utiltize the GrowthWare framework
	''' with regards to ClientChoices.
	''' </summary>
	Public Class ClientChoicesUtility
		Private m_SecurityEntityProfile As MSecurityEntityProfile = Nothing
		Private m_BClientChoices As BClientChoices = Nothing
		Private m_CachedAnonymousChoicesState As String = "AnonymousClientChoicesState"

		''' <summary>
		''' Constructor sets up the private fields
		''' </summary>
		Public Sub New()
			m_SecurityEntityProfile = SecurityEntityUtility.GetDefaultProfile()
			m_BClientChoices = New BClientChoices(m_SecurityEntityProfile, WebConfigSettings.CentralManagement)
		End Sub

		''' <summary>
		''' Returns the client choices given the account
		''' </summary>
		''' <param name="account">String</param>
		''' <returns>MClientChoicesState</returns>
		Public Function GetClientChoicesState(ByVal account) As MClientChoicesState
			Dim mRetVal As MClientChoicesState = Nothing
			If Not HttpContext.Current.Session Is Nothing Then
				mRetVal = CType(HttpContext.Current.Session(MClientChoices.SessionName), MClientChoicesState)
			End If
			If mRetVal Is Nothing Then
				If account.Trim.ToLower(CultureInfo.CurrentCulture) = "anonymous" Then
					mRetVal = CType(HttpContext.Current.Cache(m_CachedAnonymousChoicesState), MClientChoicesState)
					If mRetVal Is Nothing Then
						mRetVal = m_BClientChoices.GetClientChoicesState(account)
						CacheController.AddToCacheDependency(m_CachedAnonymousChoicesState, mRetVal)
					End If
				Else
					If mRetVal.AccountName <> account Then
						mRetVal = m_BClientChoices.GetClientChoicesState(account)
					End If
				End If
			End If
			If Not HttpContext.Current.Session Is Nothing Then
				HttpContext.Current.Session.Add(MClientChoices.SessionName, mRetVal)
			End If
			Return mRetVal
		End Function

		''' <summary>
		''' Save the client choices to the database.
		''' </summary>
		''' <param name="clientChoicesState">MClientChoicesState</param>
		''' <remarks></remarks>
		Public Sub SaveClientChoicesState(ByRef clientChoicesState As MClientChoicesState)
			m_BClientChoices.Save(clientChoicesState)
			If Not HttpContext.Current.Session Is Nothing Then
				HttpContext.Current.Items(MClientChoices.SessionName) = clientChoicesState
			End If
		End Sub
	End Class
End Namespace
